cmd/geth/simulatedbeacon/simulatedbeacon.go

```diff
@@ -0,0 +1,254 @@
+
+package simulatedbeacon
+
+import (
+	"errors"
+	"fmt"
+	"math/big"
+	"sync"
+	"time"
+
+	"github.com/ethereum/go-ethereum/beacon/engine"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/eth"
+	"github.com/ethereum/go-ethereum/eth/catalyst"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/node"
+	"github.com/ethereum/go-ethereum/rpc"
+)
+
+// withdrawalQueue implements a FIFO queue which holds withdrawals that are pending inclusion
+type withdrawalQueue struct {
+	// queued holds withdrawals that will be included in the next block
+	queued []*types.Withdrawal
+	// pending holds withdrawals that will be included in a future block
+	pending []*types.Withdrawal
+	mu      sync.Mutex
+}
+
+// getQueued returns withdrawals which are pending inclusion in the next block
+func (w *withdrawalQueue) getQueued() []*types.Withdrawal {
+	w.mu.Lock()
+	defer w.mu.Unlock()
+
+	queued := make([]*types.Withdrawal, len(w.queued))
+	copy(queued, w.queued)
+	return queued
+}
+
+// add queues a withdrawal for future inclusion
+func (w *withdrawalQueue) add(withdrawal *types.Withdrawal) error {
+	w.mu.Lock()
+	defer w.mu.Unlock()
+
+	if len(w.queued) < 10 {
+		w.queued = append(w.queued, withdrawal)
+	} else {
+		w.pending = append(w.pending, withdrawal)
+	}
+	return nil
+}
+
+// clearQueued shifts the last 10 withdrawals out of the queue
+func (w *withdrawalQueue) clearQueued() {
+	w.mu.Lock()
+	defer w.mu.Unlock()
+
+	var queueCount int
+	if len(w.pending) >= 10 {
+		queueCount = 10
+	} else {
+		queueCount = len(w.pending)
+	}
+	w.queued = make([]*types.Withdrawal, queueCount)
+	copy(w.queued, w.pending[0:queueCount])
+	w.pending = append([]*types.Withdrawal{}, w.pending[queueCount:]...)
+}
+
+type SimulatedBeacon struct {
+	shutdownCh   chan struct{}
+	eth          *eth.Ethereum
+	period       uint64
+	withdrawals  withdrawalQueue
+	feeRecipient common.Address
+	// mu gates concurrent access to the feeRecipient
+	mu sync.Mutex
+}
+
+func NewSimulatedBeacon(eth *eth.Ethereum) (*SimulatedBeacon, error) {
+	chainConfig := eth.APIBackend.ChainConfig()
+	if chainConfig.Dev == nil {
+		return nil, errors.New("incompatible pre-existing chain configuration")
+	}
+
+	return &SimulatedBeacon{
+		eth:          eth,
+		period:       chainConfig.Dev.Period,
+		withdrawals:  withdrawalQueue{[]*types.Withdrawal{}, []*types.Withdrawal{}, sync.Mutex{}},
+		feeRecipient: common.Address{},
+	}, nil
+}
+
+func (c *SimulatedBeacon) setFeeRecipient(feeRecipient *common.Address) {
+	c.mu.Lock()
+	c.feeRecipient = *feeRecipient
+	c.mu.Unlock()
+}
+
+func (c *SimulatedBeacon) getFeeRecipient() common.Address {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	return c.feeRecipient
+}
+
+// Start invokes the SimulatedBeacon life-cycle function in a goroutine
+func (c *SimulatedBeacon) Start() error {
+	c.shutdownCh = make(chan struct{})
```

Comment:
usually we initialize all channels in the constructor, it's a bit uncommon to do it here.